parflow_data_management/transport/consumers.py

```diff
@@ -0,0 +1,24 @@
+import json
+
+from channels.generic.websocket import JsonWebsocketConsumer
+
+from .tasks import generate_key_and_passphrase
+
+
+class KeyConsumer(JsonWebsocketConsumer):
+    def connect(self):
+        self.accept()
+
+    def disconnect(self, close_code):
+        pass
+
+    def receive(self, text_data):
+        if text_data == "GENERATE_KEY":
+            data = generate_key_and_passphrase()
+            self.public_key_with_passphrase(data)
```

Comment:
I think we want to initiate the request through the REST interface. The channel would just be for the notification from the server when the job is done.

Reply:
Got it. The previous implementation didn't use channels in both directions. This was a fix for the issue I came across last week. Essentially, the webpage wouldn't be updated with the passphrase and public key when the button was pressed, because that would require re-loading the webpage. This solution keeps the same websocket open the entire time.

Now that I know what the issue is, however, I can start searching for better workarounds.

Should I move the endpoint to something like `keypair/generate/`?

Reply:
So my thoughts are that the API would be similar to the other endpoints. So you would something like:

`POST /keypairs` - To trigger the generation of a key. This could return a key id, that could then be used to wait for the key message over the websocket.  
`GET /keypairs/{id}` - To get the public part of the key
`DELETE /keypairs/{id}` - To remove a key 
etc.

So basically it would match the other endpoints. The only real difference. Is the result of the key generation would come over the websocket.

Reply:
We are trying to augment the REST API with websockets ...

Reply:
I see. I imagine a user would also get the corresponding `passphrase` by doing a `GET /keypairs/{id}` with the `id` returned by the `POST`?

Reply:
We are not storing the `passphrase`, that would only ever get to the user via the websocket, we can speak in person if that would help?

Reply:
Sure.

Reply:
How does fbed1ae9434480d68fc3afc6494e0b4d16192f68 look?

Reply:
I am seeing the following error, when I try this out: 

```
raise ValueError(
ValueError: Cannot assign "<django.contrib.auth.models.AnonymousUser object at 0x7f77d7f2fb20>": "KeyPair.owner" must be a "User" instance
```

Reply:
Are you logged in?

Reply:
I am guessing not, how do I do that?

Reply:
At the main page, go to `Sign Up` at the top. Then fill out the required information.

Reply:
OK, looking good, whats left to do, run it async?

Reply:
Yes. Working through some celery + redis issues currently. Then should be good to go.